### ISSUE TITLE
Nodes never reach a status of ‘Ready’ when taints are described in the instance group spec.

### ISSUE THREAD

**Issue by ca16** [kubernetes/kops#2495]
kops version: 1.6.0-alpha.2

I’ve set up a cluster on AWS using kops. It uses Calico for networking.

When I try to create a new instance group (`kops create ig`) with taints, nodes in that instance group never reach a status of ‘Ready’ (`kubectl get nodes`). 

After poking around a bit, I think this is because `calico-node` daemon set pods are unable to get scheduled on nodes in this new instance group, because they cannot tolerate the taints I’ve added. Why I think this is the problem:

- When I run `kubectl describe node <node in new instance group>`, the message under Conditions -> Type -> Ready is `runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized`

- calico-node pods are running on every node except nodes in my new instance group (`kubectl get pods -o wide --all-namespaces`)

- if I temporarily remove the taint, or expand the toleration section in the daemon set template, calico-node pods start running on my new nodes and they reach a 'Ready' status. 

Examining the daemon set template, it looks like these pods have the following tolerations:

```
      tolerations:
      - effect: NoSchedule
        key: node-role.kubernetes.io/master
      - key: CriticalAddonsOnly
        operator: Exists
```

I'm guessing this comes from the template [here](https://github.com/kubernetes/kops/blob/master/upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template#L53).

As far as I can tell, it is not possible to specify a taint in the instance group spec and have nodes in that instance group eventually reach a status of ‘Ready’ without manually doing something else (e.g. temporarily removing the taint or updating the daemon set template through `kubectl edit`). Perhaps the toleration section in the calico-node daemon set template should catch more taints? E.g. when I look at the toleration section for the `kube-proxy` pods I find:

```
  tolerations:
  - effect: NoExecute
    operator: Exists
```

(kube-proxy pods do manage to get on to my tainted nodes)

Basically this looks like a bug to me, though if I'm mistaken about what should happen in this case please let me know!

Thanks!

**Comment by ca16** [comment]
Possible fix here:

https://github.com/ca16/kops/commit/b66b6260e79c432ec15f6165be930fd252eb563a

I'd be happy to submit a PR for this if the fix sounds good. 

Unfortunately, I haven't quite figured out how to build and test my own version of `kops` yet, so any guidance there would be much appreciated!

**Comment by willtrking** [comment]
Thanks @ca16!

Ran into the same situation with kops 1.6.0, resolved with the fix from @ca16 although for some reason I had to delete the affected instance group and recreate it after editing the calico-node daemonset.

**Comment by mtanski** [comment]
This issue also impacts knope-vxlan on kops 1.6.0,

**Comment by chrislovecnm** [comment]
This should be fixed for calico in 1.6.1 kops

**Comment by nkobber** [comment]
@chrislovecnm 
I don't think it is. Still the same [ tolerations](https://github.com/kubernetes/kops/blob/release-1.6/upup/models/cloudup/resources/addons/networking.projectcalico.org/k8s-1.6.yaml.template#L120) in the template.
Also when i upgraded kops and did a `kops update cluster`, the calico-node daemonset was updated and the tolerations i manually added were then removed.
